pkg/scheduler/internal/cache/cache.go

```diff
@@ -230,28 +253,66 @@ func (cache *schedulerCache) UpdateNodeInfoSnapshot(nodeSnapshot *nodeinfosnapsh
 	}
 
 	if len(nodeSnapshot.NodeInfoMap) > len(cache.nodes) {
-		for name := range nodeSnapshot.NodeInfoMap {
-			if _, ok := cache.nodes[name]; !ok {
-				delete(nodeSnapshot.NodeInfoMap, name)
-			}
-		}
+		cache.removeDeletedNodesFromSnapshot(nodeSnapshot)
+		updateAllLists = true
 	}
 
-	// Take a snapshot of the nodes order in the tree
+	if updateAllLists || updateNodesHavePodsWithAffinity {
+		cache.updateNodeInfoSnapshotList(nodeSnapshot, updateAllLists)
+	}
+
+	if len(nodeSnapshot.NodeInfoList) != len(nodeSnapshot.NodeInfoMap) {
+		err := fmt.Errorf("snapshot state is not consistent, length of NodeInfoList=%v not equal to length of NodeInfoMap=%v "+
+			"length of nodes in cache=%v, length of nodes in tree=%v"+
+			", trying to recover",
+			len(nodeSnapshot.NodeInfoList), len(nodeSnapshot.NodeInfoMap),
+			len(cache.nodes), cache.nodeTree.numNodes)
+		// We will try to recover by re-creating the lists for the next scheduling cycle, but still return an
+		// error to surface the problem, the error will likely cause a failure to the current scheduling cycle.
```

Comment:
"likely" or definitely?
If we can recover, I don't think we should fail the current scheduling cycle. We should log the error message and go on.

Reply:
I want to make this more visible than just a log because this is a major error.

Reply:
I don't think returning an error makes it "more visible". We just record a metric and retry.
If you think the metric makes it more visible, we can just add a metric here.

Since we can recover, I don't think we should return an error because by definition there is no error any more.

Reply:
yes, I was thinking about the failure metrics. I don't think we should update the metric here, it should be updated in one place where we test whether or not the scheduling cycle succeeded.

Reply:
If metrics is the way to make it visibly, let's create a new metric for this. Otherwise it's hard to tell what caused the current schedule_attempt error metric.

I still disagree with returning error - first there is no error any more, second it makes scheduler to waste another cycle. I think what we want here is to make this visible, and I don't think returning an error is the best way to do it.

For this type of unexpected errors, I have seen people either using a metric or have a log scan that alerts when there are too many error logs.

Reply:
> If metrics is the way to make it visibly, let's create a new metric for this. 

I don't think this is worth a separate metric, it is just another fault, do we want to create a separate metric for every fault we encounter?

> Otherwise it's hard to tell what caused the existing schedule_attempt error metric.

Admins usually look for patterns, if this is happening consistently for a specific scheduler, then an uptick in the schedule_attempt error will prompt admins to look at the logs.

> I still disagree with returning error - first there is no error any more, 

There was an error, and we are trying to surface it as best we can.

> second it makes scheduler to waste another cycle. I think what we want here is to make this visible, and I don't think returning an error is the best way to do it.

I don't see wasting a cycle in a potentially unhealthy scheduler is an issue.

> For this type of unexpected errors, I have seen people either using a metric or have a log scan that alerts when there are too many error logs.

Sure, but this is not a binary deployed in one place that we control how it is monitored. So the more signals we send the better in my opinion.

We can log an error message as well here in addition to returning the error

Reply:
I am not in favor of creating a metric. I am just arguing that the only effect of returning the error is to increment an error counter. If that's the case, why not just increment the error metric here and move on? 
If we retry, the pod is added back to the queue and has to wait until it's picked up again (which may take some time). I think it's not "fair" for the pod.

It's weird to me that an error is properly handled and then returned. I think we should log the error.

Reply:
> If that's the case, why not just increment the error metric here and move on?

This was addressed above, do we want to increment a failure metric everywhere we fail, or just in one place? I think just in one place.

> If we retry, the pod is added back to the queue and has to wait until it's picked up again (which may take some time). I think it's not "fair" for the pod.

This is a weird "fair" argument :)

> 
> It's weird to me that an error is properly handled and then returned. I think we should log the error.

It is not guaranteed that we actually recovered. Both options have pros and cons, there is clearly no optimal solution here.

Reply:
> This was addressed above, do we want to increment a failure metric everywhere we fail, or just in one place? I think just in one place.

This is different. I actually think this error can be properly recovered. If I am wrong, then yeah let's do a best effort recovery and return the error.

> It is not guaranteed that we actually recovered. Both options have pros and cons, there is clearly no optimal solution here. I added a log message.

I believe by recreating the snapshot we can actually guarantee a proper recover. If this is not the case, yeah then I agree to return the error.

Reply:
SG, then as I mentioned in the other comment, we don't have the logic to properly recover, so we can keep this as is, and perhaps in a separate PR we can try to come up with an approach to test for inconsistent state in the map and properly recover.